### PR TITLE
Add query specifications

### PR DESCRIPTION
**Pull request by Grasmus**

### CODE REVIEW

**Review by Eugene-pas — APPROVED** [review]

**Review by Eugene-pas — CHANGES_REQUESTED** [review]
Testing this push

File: Core/Interfaces/IRepository.cs [hunk 1]
```diff
@@ -15,5 +15,6 @@ public interface IRepository<TEntity> where TEntity : class
         Task UpdateAsync(TEntity entityToUpdate);
         Task<int> SaveChangesAsync();
         IQueryable<TEntity> Query(params Expression<Func<TEntity, object>>[] includes);
+        IEnumerable<TEntity> FindWithSpecification(ISpecification<TEntity> specification);
```
Comment (vladtymo) [inline Core/Interfaces/IRepository.cs:21]: Why don't you use the async method form?

File: Core/Services/AuthenticationService.cs [hunk 2]
```diff
@@ -118,7 +123,17 @@ private async Task<RefreshToken> CreateRefreshToken(string userId)
 
         public async Task<UserAutorizationDTO> RefreshTokenAsync(UserAutorizationDTO userTokensDTO)
         {
-            var refreshToken = GetRefreshToken(userTokensDTO.RefreshToken);
+<<<<<<< HEAD
+            var refreshToken = (await _refreshTokenRepository
+                .FindWithSpecificationAsync(new GetRefreshTokenByToken(userTokensDTO.RefreshToken)))
+=======
+            var refreshToken = _refreshTokenRepository
+                .FindWithSpecification(new GetRefreshTokenByToken(userTokensDTO.RefreshToken))
```
Comment (YevheniiVovk) [inline Core/Services/AuthenticationService.cs:131]: Make method that return First element. Please DO NOT select all items by criteria and take first element from memory

File: Core/Interfaces/CustomService/IRoleService.cs [hunk 3]
```diff
@@ -1,10 +1,11 @@
...
+using System.Threading.Tasks;
 
 namespace Core.Interfaces.CustomService
 {
     public interface IRoleService
     {
-        IdentityRole GetIdentityRoleByName(string roleName);
-        int GetRoleByName(string name);
+        Task<IdentityRole> GetIdentityRoleByNameAsync(string roleName);
+        Task<int> GetRoleByNameAsync(string name);
```
Comment (YevheniiVovk) [inline Core/Interfaces/CustomService/IRoleService.cs:9]: Already exist 
https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.identity.rolemanager-1?view=aspnetcore-6.0

File: Core/Interfaces/ISpecification.cs [hunk 4]
```diff
@@ -0,0 +1,14 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq.Expressions;
+
+namespace Core.Interfaces
+{
+    public interface ISpecification<TEntity>
+    {
+        Expression<Func<TEntity, bool>> Criteria { get; }
+        IList<Expression<Func<TEntity, object>>> Includes { get; }
```
Comment (YevheniiVovk) [inline Core/Interfaces/ISpecification.cs:10]: Why from Object ?
TEntity, object

File: Core/Services/GoodCategoryService.cs [hunk 5]
```diff
@@ -31,10 +30,19 @@ public async Task<GoodCategoryListDTO> GetAllGoodCategoryAsync()
                 {GoodCategories = _mapper.ProjectTo<GoodCategoryDTO>(list.AsQueryable())};
         }
 
-        public int GetGoodCategoryByName(string goodCategoryName)
+        public async Task<int> GetGoodCategoryByNameAsync(string goodCategoryName)
         {
-            var goodCategory =  _goodCategoryRepository.Query().FirstOrDefault(goodCategory => goodCategory.Name == goodCategoryName.ToUpper());
+<<<<<<< HEAD
```
Comment (YevheniiVovk) [inline Core/Services/GoodCategoryService.cs:35]: What is it ?

File: Core/Services/GoodCategoryService.cs [hunk 6]
```diff
@@ -31,10 +30,19 @@ public async Task<GoodCategoryListDTO> GetAllGoodCategoryAsync()
...
+            var goodCategory = (await _goodCategoryRepository
+                .FindWithSpecificationAsync(new GetGoodCategoryByName(goodCategoryName)))
+=======
+            var goodCategory =  _goodCategoryRepository
+                .FindWithSpecification(new GetGoodCategoryByName(goodCategoryName))
+>>>>>>> 8a83ac67acfb270be6b52f5d2140f970db84dace
```
Comment (YevheniiVovk) [inline Core/Services/GoodCategoryService.cs:41]: Same what is this ?

File: Core/Services/OfferService.cs [hunk 7]
```diff
@@ -44,9 +44,9 @@ public async Task<OfferCreateDTO> CreateOfferAsync(OfferCreateDTO offerCreate, s
             offer.OfferCreatorId = userId;
             offer.CreationDate = DateTimeOffset.UtcNow;
             offer.IsClosed = false;
-            offer.CreatorRoleId = _roleRepository.GetRoleByName(offerCreate.Role);
-            offer.GoodCategoryId = _goodCategoryRepository
-                .GetGoodCategoryByName(offerCreate.GoodCategory);
+            offer.CreatorRoleId = await _roleRepository.GetRoleByNameAsync(offerCreate.Role);
```
Comment (YevheniiVovk) [inline Core/Services/OfferService.cs:47]: Please use [RoleManager ](https://docs.microsoft.com/en-us/dotnet/api/microsoft.aspnetcore.identity.rolemanager-1?view=aspnetcore-6.0) or explain why it should not be used

File: Core/Services/RoleService.cs [hunk 8]
```diff
@@ -4,6 +4,11 @@
 using Core.Interfaces;
 using Core.Interfaces.CustomService;
 using System.Linq;
+using Core.Specifications;
+<<<<<<< HEAD
```
Comment (YevheniiVovk) [inline Core/Services/RoleService.cs:8]: What is it?

File: Core/Services/RoleService.cs [hunk 9]
```diff
@@ -4,6 +4,11 @@
...
+using System.Threading.Tasks;
+=======
+>>>>>>> 8a83ac67acfb270be6b52f5d2140f970db84dace
```
Comment (YevheniiVovk) [inline Core/Services/RoleService.cs:11]: Same here

**Comment by YevheniiVovk** [comment]
Please read about git and conflict's and before request the review of code make sure that code can be run and works !

**Pull request closed without merging**